EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,263 @@
+
+<img src="https://github.com/0xProject/EIPs/tree/master/EIPS/eip-eth_signTypedData/personal_sign.png"  width="500px">
+
+The main problem is that signers don't have enough metadata to display the DApp's real intent effectively. In order to do so, they need the plaintext input that the dApp wishes hashed and signed by the user.
+
+## Specification
+
+This EIP proposes a new JSON RPC method to the `eth` namespace: `eth_signTypedData`. It accepts an array of values together with their specified type and human-readable name. The [json-schema](http://json-schema.org/) draft is defined below.
+
+### Params JSON Schema:
+```json-schema
+{
+  items: {
+    properties: {
+      name: {type: 'string'},
+      type: {type: 'string'}, // Solidity type as described here: https://github.com/ethereum/solidity/blob/93b1cc97022aa01e7daa9816bcc23108bbe008b5/libsolidity/ast/Types.cpp#L182
+      value: {
+        oneOf: [
+          {type: 'string'},
+          {type: 'number'},
+          {type: 'boolean'},
+        ],
+      },
+    },
+    type: 'object',
+  },
+  type: 'array',
+}
+```
+
+### Example params:
+// For this state channel POC: https://medium.com/@matthewdif/ethereum-payment-channel-in-50-lines-of-code-a94fad2704bc
+```javascript
+typedData = [
+    {
+      "name": "channel",
+      "type": "address",
+      "value": "0xb088a3Bc93F71b4DE97b9De773e9647645983688",
+    },
+    {
+      "name": "value",
+      "type": "uint",
+      "value": 42,
+    },
+];
+```
+
+### How it can look in signer UI:
+
+<img src="https://github.com/0xProject/EIPs/tree/master/EIPS/eip-eth_signTypedData/eth_signTypedData.png" width="500px">
+
+
+It's important to make the schema part of the signature (explanation can be found in “Rationale” section). The way the schema will be combined with the values to generate the hash signed by the user is shown below. First, the schema is encoded into a string using a method similar to [solidity events signatures](http://solidity.readthedocs.io/en/develop/contracts.html#low-level-interface-to-logs). Then it's hash is prepended to the data array before hashing it.
+
+### Pseudocode examples:
+
+```javascript
+// Client-side code example
+const typedData = [
+  {
+    'type': 'string',
+    'name': 'message',
+    'value': 'Hi, Alice!',
+  },
+  {
+    'type': 'uint',
+    'name': 'value',
+    'value': 42,
+  },
+];
+const signature = await web3.eth.signTypedData(signerAddress, typedData);
+```
+
+```javascript
+// Signer code JS example
+import * as _ from 'lodash';
+import * as ethAbi from 'ethereumjs-abi';
+
+const schema = _.map(typedData, entry => `${entry.type} ${entry.name}`).join(',');
+const schemaHash = ethAbi.soliditySHA3(['string'], schema);
+const data = _.map(typedData, 'value');
+const types = _.map(typedData, 'type');
+const hash = ethAbi.soliditySHA3(
+  ['bytes32', ...types],
```

Comment:
This is not consistent with the currrent state of Metamask's implementation, nor the first post of this PR.